std/traits.d

```diff
@@ -2658,6 +2659,150 @@ template RepresentationTypeTuple(T)
     static assert(R2.length == 2 && is(R2[0] == int) && is(R2[1] == immutable(Object)));
 }
 
+// Layout
+/**
+Returns the layout in memory of type `T` as an `AliasSeq`. For each field in
+`T`, the result includes the offset of the field followed by the type of the
+field. Detailed information depends on the type's kind, as follows:
+
+$(UL
+$(LI If `T` is a `struct` or `union` type, then `Layout!T` is an `AliasSeq`
+containing each field preceded by its offset. Fields that are in turn of
+`struct` or `union` types are decomposed transitively. The result is ordered
+in increasing offset order. (Two or more fields may have the same offset due to
+`union`s.) For a given offset, types are in the order of the declaration
+they are a part of.)
+
+$(LI If `T` is a `class` or `interface` type, the layout is similar to the one
+return for `struct`s, with one difference: classes and interfaces have
+additional opaque data managed by the compiler. Usually these are placed at the
+front of the object. This data is represented in the layout as pointers to opaque
+types. The presence, number, and types of additional fields are
+implementation-dependent and should not be relied upon in portable code. All these
+additional types have names that start with a double underscore.)
+
+$(LI If `T` is a statically-sized array, the layout is an iteration of the
+layouts of each field of the array. Note that this may create large tuples.)
+
+$(LI For all other types, the layout is $(D AliasSeq!(size_t(0), T)).)
+
+)
+ */
+template Layout(T)
+{
+    private template Add(size_t offset, U...)
+    {
+        static if (!U.length)
+            alias Add = AliasSeq!();
+        else
+            alias Add = AliasSeq!(U[0] + offset, U[1],
+                                     Add!(offset, U[2 .. $]));
+    }
+
+    static if (!is(T == Unqual!T))
+    {
+        alias Layout = Layout!(Unqual!T);
+    }
+    else static if (is(T == U[n], U, size_t n))
+    {
+        static if (n == 0)
+            alias Layout = AliasSeq!();
+        else static if (is(U == class) || is(U == interface))
+            alias Layout = AliasSeq!(size_t(0), U, Add!(U.sizeof, Layout!(U[n - 1])));
+        else
+            alias Layout = AliasSeq!(Layout!U, Add!(U.sizeof, Layout!(U[n - 1])));
+    }
+    else static if (!is(T == struct) && !is(T == union)
+                    && !is(T == class) && !is(T == interface))
```

Comment:
Instead of all the nots, make the body of this the `else` clause which is case 4 "for all other types".

Reply:
I respectfully disagree here. When getting into code, I found this pattern:

```d
if (expression)
{
    ... one-liner handling the simple case ...
}
else
{
    ... 70 lines handling the complex case ...
}
```

considerably easier to get into than the converse:

```d
if (expression)
{
    ... 70 lines handling the complex case ...
}
else
{
    ... one-liner handling the simple case ...
}
```

This is because whilst reading the 70 lines the other branch, also potentially as long, is still looming. If the simple case is eliminated early, I can understand it easily and then mentally dispense with it.

Also: 

* I find recursions easier to read if the limit condition is placed first
* I find loops with filtering "of all elements just process those fulfilling some condition..." easier to read if they eliminate the mismatches with `if (mismatch) continue;`

I do understand others believe differently! Live and let live is my principle.